### PR TITLE
Temporary files evict fs cache

### LINKED ISSUE #42158: Temporary data on disk should cooperate with the filesystem cache.
**Use case**

Local cache and temporary data on disk are reusing the same space and should work and be limited together.

**Describe the solution you'd like**

If the path for temporary data on disk is identical to the filesystem cache path, the temporary data on the disk will use the cache object to allocate space.

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Temporary data (for external sorting, aggregation, and JOINs) can share storage with the filesystem cache for remote disks and evict it, close #42158

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/Cache/FileSegment.cpp [hunk 1]
```diff
@@ -99,6 +99,7 @@ size_t FileSegment::getFirstNonDownloadedOffsetUnlocked(std::unique_lock<std::mu
     return range().left + getDownloadedSizeUnlocked(segment_lock);
 }
 
+/// same as getFirstNonDownloadedOffset ?
```
Comment (kssenii) [inline src/Interpreters/Cache/FileSegment.cpp:102]: Yes, it is the same with current code, but it will not be the same after https://github.com/ClickHouse/ClickHouse/pull/40026 is merged. I extracted refactoring of cache code into another pr and merged it some time ago because of the number of conflicts arising all the time, so currently it might be a little confusing that there are two similar methods.

**Comment by vdimir** [comment]
Lots of failures in CI, seems it took bad commit https://github.com/ClickHouse/ClickHouse/commit/b784128c888e870309a530c90a4dedf684e79453

But `Unit tests` looks related, will fix it with further changes

File: src/Disks/IO/FileCachePlaceholder.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                                                                 /* cache_log_ */ nullptr,
+                                                                 /* query_id_ */ "",
+                                                                 /* source_path_ */ key);
+
+    /* Sometimes several calls of tryReserve are required,
+     * because it allocates space in current segment and its size may be less than requested_size.
+     * If we can't write any data, we need to free all space occupied by current reservation.
+     * To do this we will just destroy current cache_writer.
+     */
+    while (requested_size > 0)
```
Comment (vdimir) [inline src/Disks/IO/FileCachePlaceholder.cpp:60]: Move this loop inside `FileSegmentRangeWriter`, so `tryWrite/tryReserve` would allocate as much segments as needed

File: src/Disks/IO/FileCachePlaceholder.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        size_t current_offset = cache_writer->currentOffset();
+        size_t written = cache_writer->tryReserve(requested_size, current_offset, /* is_persistent */ false, /* strict */ false);
+        if (written == 0)
+        {
+            cache_writer->finalize(/* clear */ true);
```
Comment (vdimir) [inline src/Disks/IO/FileCachePlaceholder.cpp:66]: Set some `is_temporary` flag to file segments so they will remove themself from the cache in their destructors and calling `finalize(/* clear */ true)` will be unnecessary

**Review by vdimir — COMMENTED** [review]
Some comments after discussion with @kssenii

**Comment by vdimir** [comment]
CI failures (@ https://github.com/ClickHouse/ClickHouse/pull/43457/commits/48867e5cbb16deb69960541c908a40ee85e0a568) with links to known issues:

> Stateless tests (tsan) [4/5] — fail: 1, passed: 919, skipped: 10 [Details](https://s3.amazonaws.com/clickhouse-test-reports/43457/48867e5cbb16deb69960541c908a40ee85e0a568/stateless_tests__tsan__[4/5].html)

`01072_window_view_multiple_columns_groupby` timeout, https://github.com/ClickHouse/ClickHouse/issues/43659


> Stress test (asan) — Cannot start clickhouse-server [Details](https://s3.amazonaws.com/clickhouse-test-reports/43457/48867e5cbb16deb69960541c908a40ee85e0a568/stress_test__asan_.html)

```Part ... was found on disk `local_cache` which is not defined in the storage policy```, https://github.com/ClickHouse/ClickHouse/issues/41218


> Stress test (tsan) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt) [Details](https://s3.amazonaws.com/clickhouse-test-reports/43457/48867e5cbb16deb69960541c908a40ee85e0a568/stress_test__tsan_.html)

```MetadataStorageFromDiskTransaction::commit(): Code: 566. DB::ErrnoException: Cannot rmdir```, https://github.com/ClickHouse/ClickHouse/issues/41369

> Stress test (ubsan) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt) [Details](https://s3.amazonaws.com/clickhouse-test-reports/43457/48867e5cbb16deb69960541c908a40ee85e0a568/stress_test__ubsan_.html)

```The local set of parts of table ... doesn't look like the set of parts in ZooKeeper```, https://github.com/ClickHouse/ClickHouse/issues/37664

---

@kssenii If you don't mind, I'll merge this PR (right after fixing the conflict) since CI is ok. If you have any comments, I'll fix them in a follow-up PRs.
